fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMutuallyByRecords.java

```diff
@@ -0,0 +1,529 @@
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    private static Range partlyUnBuiltRange(List<Range> missingRanges, Range fragmentRange) {
+        // Return null if the range appears to be fully built in the missing ranges list. Else
+        // return the un-built range of (fragmentRange ∩ missingRanges)
+        // * Assuming that the missing ranges are sorted and squashed
+        // * Less optimized than the fullyUnBuiltRanges (this function is expected to be called only a handful
+        //   of times at the end of the indexing process)
+        for (Range range: missingRanges) {
+            if (ByteArrayUtil.compareUnsigned(range.begin, fragmentRange.end) > 0 ||
+                    ByteArrayUtil.compareUnsigned(range.end, fragmentRange.begin) < 0) {
+                // no range overlap
+                continue;
+            }
+            // Here: some range overlap, shrink to fragmentRange ∩ range
+            byte[] begin = ByteArrayUtil.compareUnsigned(range.begin, fragmentRange.begin) >= 0 ?
+                           range.begin : fragmentRange.begin;
+            byte[] end = ByteArrayUtil.compareUnsigned(range.end, fragmentRange.end) <= 0 ?
+                         range.end : fragmentRange.end;
+            if (ByteArrayUtil.compareUnsigned(begin, end) < 0) {
+                // if non-empty range
+                return notEmptyRange(new Range(begin, end));
+            }
+        }
+        return null;
+    }
+
+    @Nullable
+    private static Range notEmptyRange(Range range) {
+        // check the (rare, maybe test only) case of empty ranges
+        return ByteArrayUtil.compareUnsigned(range.begin, range.end) >= 0 ?
+               null : range;
+    }
+
+    List<Range> sortAndSquash(List<Range> ranges) {
+        ranges.sort((a, b) -> ByteArrayUtil.compareUnsigned(a.begin, b.begin));
+        boolean squasshed = false;
+        for (int i = 0; i < ranges.size() - 1; i++) {
+            if (ByteArrayUtil.compareUnsigned(ranges.get(i).end, ranges.get(i + 1).begin) >= 0) {
+                squasshed = true;
+                ranges.set(i + 1, new Range(ranges.get(i).begin, ranges.get(i + 1).end));
+                ranges.set(i, null);
+            }
+        }
+        return squasshed ?
+               ranges.stream().filter(Objects::nonNull).collect(Collectors.toList()) :
+               ranges;
+    }
+
+    private static CompletableFuture<Void> insertRanges(Transaction transaction,
+                                                        List<RangeSet> rangeSets,
+                                                        byte[] start, byte[] end) {
+        return AsyncUtil.whenAll(rangeSets.stream().map(set -> set.insertRange(transaction, start, end, true)).collect(Collectors.toList()));
+    }
+
+    private void infiniteLoopProtection(final Range range, final List<Range> missingRanges) {
+        String token = range.toString();
+        if (token.equals(loopProtectionToken)) {
```

Comment:
If it iterates between ranges, it can only reach 3rd iteration. 
(As I've told Alec) this protection is needed because when the algorithm finds a candidate range to build, it calls `iterateAllRanges` without advancing the range pointer. Since `iterateAllRanges` is external to this module, this call is suspected as a no-op that creates an infinite loop.

Reply:
ok, I think I understand better.
If `infiniteLoopProtection` is called with `rangeA`, you trust two things:
1. if rangeA is built, `rangeA` will be removed from the `missingRanges`
2. if `rangeA` is not built, that's the range that will be selected on the second loop, because it will be in the same spot in `missingRanges`

If `fullyUnbuiltRange` or `partlyUnbuiltRange` were non-deterministic then this check would not catch that, but that's not a concern here.